python/ray/tune/suggest/hebo.py

```diff
@@ -0,0 +1,359 @@
+            points_to_evaluate: Optional[List[Dict]] = None,
+            evaluated_rewards: Optional[List] = None,
+            random_state_seed: Optional[int] = None,
+            **kwargs):
+        assert hebo is not None, (
+            "HEBO must be installed!. You can install HEBO with"
+            " the command: `pip install "
+            "git+https://github.com/huawei-noah/noah-research.git"
+            "#subdirectory=HEBO`.")
+        if mode:
+            assert mode in ["min", "max"], "`mode` must be 'min' or 'max'."
+        if random_state_seed:
+            assert random_state_seed is None or isinstance(
+                random_state_seed, int
+            ), "random_state_seed must be None or int, got '{}'.".format(
+                type(random_state_seed))
+        super(HEBOSearch, self).__init__(metric=metric, mode=mode)
+
+        if isinstance(space, dict) and space:
+            resolved_vars, domain_vars, grid_vars = parse_spec_vars(space)
+            if resolved_vars:
+                raise TypeError(SPACE_ERROR_MESSAGE)
+            if domain_vars or grid_vars:
+                logger.warning(
+                    UNRESOLVED_SEARCH_SPACE.format(
+                        par="space", cls=type(self)))
+                space = self.convert_search_space(space)
+        elif space is not None and not isinstance(
+                space, hebo.design_space.design_space.DesignSpace):
+            raise TypeError(SPACE_ERROR_MESSAGE +
+                            " Got {}.".format(type(space)))
+
+        self._hebo_config = kwargs
+        self._random_state_seed = random_state_seed
+        self._space = space
+        self._points_to_evaluate = points_to_evaluate
+        self._evaluated_rewards = evaluated_rewards
+        self._initial_points = []
+        self._live_trial_mapping = {}
+
+        self._opt = None
+        if space:
+            self._setup_optimizer()
+
+    def _setup_optimizer(self):
+        # HEBO internally minimizes, so "max" => -1
+        if self._mode == "max":
+            self._metric_op = -1.
+        elif self._mode == "min":
+            self._metric_op = 1.
+
+        if self._metric is None and self._mode:
+            # If only a mode was passed, use anonymous metric
+            self._metric = DEFAULT_METRIC
+
+        if self._random_state_seed is not None:
+            np.random.seed(self._random_state_seed)
+            torch.random.manual_seed(self._random_state_seed)
+
+        self._opt = hebo.optimizers.hebo.HEBO(
+            space=self._space, **self._hebo_config)
+
+        if self._points_to_evaluate:
+            _validate_warmstart(self._space.para_names,
+                                self._points_to_evaluate,
+                                self._evaluated_rewards)
+            if self._evaluated_rewards:
+                self._opt.observe(
+                    pd.DataFrame(self._points_to_evaluate),
+                    np.array(self._evaluated_rewards))
```

Comment:
So we validate that the lengths are equal in `_validate_warmstart`. I was wondering if we should allow `len(self_points_to_evaluate) > len(self._evaluated_rewards)`, so that for the first `len(self._evaluated_rewards)` we would observe the points_to_evaluate and the remaining points will be then the next suggestions. What do you think?

Reply:
I think that would quickly become confusing to users. I can imagine a situation where the indices in lists don't match my mistake because something is added to the front of `points_to_evaluate` and such.

TBH I don't think the current two list approach is a good one. I feel like it would be better to have one list of tuples `(points, reward)`, and allow for `reward=None`. That way the chance of indices being mismatched would be removed. However changing that in other searchers would be an API breaking change.

Reply:
Let's leave it like this for now then, i.e. require the same length if both are given. If we opt to change it in the future we will have to find a backwards compatible way anyway.